### DISCUSSION TITLE
What is the size of the largest single file supported?

### DISCUSSION THREAD

**Discussion by paochiang** [seaweedfs/seaweedfs#2124] [Q&A]
What is the size of the largest single file supported? 

I tried to upload one 1TB `.tar` file through `weed mount`,  no error  happened, but I can't see the file after writing is completed.

`weed version`: version 30GB 2.48 45a7622 linux amd64

**Comment by prologic** [comment]
How did you form your cluster out of curiosity? I've been able to write ~10M, ~100M and ~1GB files. But unfortunately I run into I/O errors reading them back which I haven't. yet been able to pint-point and repdoruce.

**Reply by paochiang** [reply]
I setup seaweedfs in the K8s cluster(total 8 nodes).

- 3 x seaweedfs master
- 3 x seaweedfs filer
- 8 x seaweedfs volume (each node has 15TB disk that was configured for storage of seaweedfs volume)

**Comment by chrislusf** [comment]
For mount, the new file is written to local disk first, then upload.

Do you have enough disk space for 1TB?

**Reply by paochiang** [reply]
> Do you have enough disk space for 1TB?

yes, disk space is enough.

> For mount, the new file is written to local disk first, then upload.

Is it uploaded after the entire file is written to local disk? What is the path of local disk?
I `weed mount` in a sftp server,  then i upload 1TB file to `weed mount` target dir through sftp client.

![image](https://user-images.githubusercontent.com/28077875/121633926-4bd45f80-cab6-11eb-9521-70031315acc5.png)

sftp client write 100%, and no error hanppend,  but `ls` cannot see the file. I also enter into `weed shell`, and cannot see file's metadata.

**Reply by chrislusf** [reply]
need to see the `weed mount` logs. Better if you run it as `weed -v=4 mount ...`

**Reply by paochiang** [reply]
can not get the `weed mount` logs now.
I will reproduce it with `weed -v=4 mount ...`,  but this will spend a lot of time. After I get the results, I will show you the log.

**Reply by prologic** [reply]
> For mount, the new file is written to local disk first, then upload.

This limits the total size of files you can write to a SeaweedFS cluster right? Because even if your cluster has >1TB of capacity you need 1TB of local disk space just so the write and background syncs can happen right?

I can only assume this design is for performance reasons? Writing directly to remote storage is a lot slower than writing to la local file system first.

Is this local read/write some kind of cache? Or can you expect to read/write to the volume directly on the local file system in addition to the `weed mount`?

**Reply by chrislusf** [reply]
This is to better align the file content. Sometimes the OS may send write content with overlapping intervals, especially random writes. Directly sending it to the cluster will cause too many overwritten chunks.

https://github.com/chrislusf/seaweedfs/blob/master/weed/filesys/filehandle.go#L42
You can toggle this line to try with the previous all in memory algorithm, or maybe help to improve current one.

**Reply by prologic** [reply]
Thank you, I'll take a look. I guess it's a trade-off between memory usage and disk usage here?

**Reply by paochiang** [reply]
> can not get the `weed mount` logs now.
> I will reproduce it with `weed -v=4 mount ...`, but this will spend a lot of time. After I get the results, I will show you the log.

Sorry, I have no machine to do this test now,  
and need to wait a few days for idle machine to reproduce it.

**Reply by chrislusf** [reply]
it's ok to wait a bit.

What use case needs to copy 1TB file? I would think most cases the large files are already split.

**Reply by paochiang** [reply]
> What use case needs to copy 1TB file? I would think most cases the large files are already split.

yes, it isn't actual application scenario, just a test scenario.

**Comment by nagy** [comment]
Maybe for the case of this large file you could use `weed filer.copy`. I did a quick test locally did not see any memory usage increase with a large file. So:

```sh
$ weed filer.copy oneTB.bin http://localhost:8888/path/to/folder/
```

**Reply by paochiang** [reply]
tks.
For some reason, I must use SFTP tool to upload file, so what I can do is to `wee mount` the target path to the SFTP's target path in SFTP server.

**Comment by paochiang** [comment]
I tried to reproduce it,  but successfully uploaded 100GB、500GB and 1TB files.

So, I have further communicated my colleagues who feedback this question.  

Finally, it was found that he terminated the transmission when the SFTP client displayed 100% is completed（In fact, at this time just successfully uploaded the file to the local disk: `/tmp/sw/xxxxxx`，and file is being uploaded to seaweedfs, it will spend more time than SFTP transfer from 0% to 100%).

So, upload file failed is our own reasons.

But，I found that if upload file failed, the temporary file in the directory: `/tmp/sw` will not be deleted, need to clean up these files manually（I think that these temporary files should be automatically cleaned regardless of transmission success or failure）.
